src/linear_algebra/tensor_algebra.lean

```diff
@@ -0,0 +1,378 @@
+-- of the tensor algebra.
+@[nolint doc_blame]
+inductive pre
+| of : M → pre
+| zero : pre
+| one : pre
+| mul : pre → pre → pre
+| add : pre → pre → pre
+| smul : R → pre → pre
+
+
+namespace pre
+
+instance : inhabited (pre R M) := ⟨pre.zero⟩
+
+-- These instances are only used to simplify the notation.
+@[nolint doc_blame]
+def has_coe : has_coe M (pre R M) := ⟨of⟩
+@[nolint doc_blame]
+def has_mul : has_mul (pre R M) := ⟨mul⟩
+@[nolint doc_blame]
+def has_add : has_add (pre R M) := ⟨add⟩
+@[nolint doc_blame]
+def has_zero : has_zero (pre R M) := ⟨zero⟩
+@[nolint doc_blame]
+def has_one : has_one (pre R M) := ⟨one⟩
+@[nolint doc_blame]
+def has_scalar : has_scalar R (pre R M) := ⟨smul⟩
+@[nolint doc_blame]
+def has_neg : has_neg (pre R M) := ⟨smul (-1 : R)⟩
+
+end pre
+
+local attribute [instance] pre.has_coe
+local attribute [instance] pre.has_mul
+local attribute [instance] pre.has_add
+local attribute [instance] pre.has_zero
+local attribute [instance] pre.has_one
+local attribute [instance] pre.has_scalar
+local attribute [instance] pre.has_neg
+
+@[nolint doc_blame]
+def lift_fun {A : Type u3} [ring A] [algebra R A] (f : M →ₗ[R] A) : pre R M → A :=
+  λ t, pre.rec_on t f 0 1 (λ _ _, (*)) (λ _ _, (+)) (λ x _ a, x • a)
+
+@[nolint doc_blame]
+inductive rel : (pre R M) → (pre R M) → Prop
+-- force of to be linear
+| add_lin {a b : M} : rel ↑(a + b) (↑a + ↑b)
+| smul_lin {r : R} {a : M} : rel ↑(r • a) (r • ↑a)
+| zero_lin : rel ↑(0 : M) 0
+-- add gives a commutative group
+| add_assoc {a b c : pre R M} : rel (a + b + c) (a + (b + c))
+| add_comm {a b : pre R M} : rel (a + b) (b + a)
+| neg_add {a : pre R M} : rel (-a + a) 0
+| zero_add {a : pre R M} : rel (0 + a) a
+-- mul gives a monoid
+| mul_assoc {a b c : pre R M} : rel (a * b * c) (a * (b * c))
+| one_mul {a : pre R M} : rel (1 * a) a
+| mul_one {a : pre R M} : rel (a * 1) a
+-- distributivity
+| left_distrib {a b c : pre R M} : rel (a * (b + c)) (a * b + a * c)
+| right_distrib {a b c : pre R M} : rel ((a + b) * c) (a * c + b * c)
+-- algebra structure
+| map_one : rel ((1 : R) • (1 : pre R M)) 1
+| map_mul {r s : R} : rel ((r * s) • 1) ((r • (1 : pre R M)) * (s • 1))
+| map_add {r s : R} : rel ((r + s) • (1 : pre R M)) (r • 1 + s • 1)
+| map_zero : rel ((0 : R) • (1 : pre R M)) 0
+| commutes {r : R} {a : pre R M} : rel ((r • (1 : pre R M)) * a) (a * (r • 1))
+| smul_def {r : R} {a : pre R M} : rel (r • a) ((r • 1) * a)
+-- compatibility
+| mul_compat_left {a b c : pre R M} : rel a b → rel (a * c) (b * c)
+| mul_compat_right {a b c : pre R M} : rel a b → rel (c * a) (c * b)
+| add_compat_left {a b c : pre R M} : rel a b → rel (a + c) (b + c)
+| add_compat_right {a b c : pre R M} : rel a b → rel (c + a) (c + b)
+| smul_compat {r : R} {a b : pre R M} : rel a b → rel (r • a) (r • b)
+
+end tensor_algebra
+
+/--
+The tensor algebra of the module `M` over the commutative ring `R`.
+-/
+def tensor_algebra := quot (tensor_algebra.rel R M)
```

Comment:
`tensor_product` uses something more like `(add_con_gen (tensor_algebra.rel R M)).quotient` here - what's the tradeoff between the two approaches?

Reply:
As I understand it, `add_con_gen` is essentially used in an additive context to generate, given `rel`, the smallest equivalence relation containing `rel` and which is compatible with addition. This is more complicated since we're defining a ring (in fact, an algebra!) structure. 

Note that `rel` is not an equivalence relation, so I used `quot` instead of `quotient`. I could have used `eqv_gen`, but I couldn't see any benefit in doing this -- it would have just added more typing.

Reply:
Is there any reason why we would want to actually have a setoid here?